### ISSUE TITLE
APScheduler Usage in a web-app. Multiple jobs execution on uwsgi

### ISSUE THREAD

**Issue by saadqc** [agronholm/apscheduler#160]
Hi,
I am using apscheduler on production. So far it was working best until i face issue of jobs being executed multiple times. The issue is that in my app uwsgi app creates 5 process. And in each process a scheduler instance is initialized and running. So, when job time comes, every instance tries to run that job.
So, we cannot use apscheduler in multi process server

The best solution can be to use apscheduler callback as a separate process in a worker. 
This way when a job time comes, job will be executed in that separate worker instead of flask app instance just like celery.

http://stackoverflow.com/questions/24352634/would-starting-apscheduler-in-a-uwsgi-app-end-up-with-one-scheduler-for-each-wor

http://stackoverflow.com/questions/16053364/make-sure-only-one-worker-launches-the-apscheduler-event-in-a-pyramid-web-app-ru

**Comment by agronholm** [comment]
It's not clear what this issue wants me to do. Is it a bug report or a feature request? If the latter, then exactly what feature is being requested?

**Comment by agronholm** [comment]
My recommendation is to start apscheduler as a separate process and connect to it via some RPC mechanism to add jobs.

**Comment by saadqc** [comment]
@agronholm Actually, its a feature. apscheduler, currently, is not compatible to use with any server or web-app because it runs inside flask/django/pyramid app. So, when multiple apps are created on server side, multiple apscheduler instances are also initialized with it. 

So, my point is that the callback logic can be moved to a separate independent process like celery workers.

--> python apscheduler_callback.py

in the worker method

```
def callback_job(*args, **kwargs):
      # do stuff

```

this way the issue can be cater well and we don't need to communicate from apscheduler and web-app manually using pipe as it needs a lot of implementation.

**Comment by agronholm** [comment]
I still don't understand what you're asking for. Could you rephrase?

**Comment by saadqc** [comment]
"My recommendation is to start apscheduler as a separate process and connect to it via some RPC mechanism to add jobs."

There are five process created by uwsgi each running a flask app along with a apscheduler instance.
So, there are total 5 different scheduler instances.

When we send a request to endpoint, uwsgi passes request to one of the process.
In that endpoint we can add job/remove/modify/pause/resume/get job. 
But apscheduler polls with redis(jobstore) is to check if a job time is due. In the current scenario, in all five processes, apscheduler instance  invoke callback. So, instead of one time job execution, 5 jobs are executed.

Also, you already answered the questions i shared on stackoverflow.

Feature:

If we add the apscheduler job callback in a separate file and that file is executed as a separate apscheduler process.

Then we can have

--> a web-app where multiple scheduler instances are created. Where we can create/delete/add/get operations.

--> a worker process where a single scheduler instance is created. Which is just for executing job when its time come. (callback of any job)

**Comment by saadqc** [comment]
this pattern is production ready and similar to celery. This way apscheduler can be integrated with any web-app framework.

**Comment by agronholm** [comment]
If you have workers and each APScheduler instance creates an external process, then you will end up with 5 processes. Instead, you'll want to start an external process that runs a single scheduler instance and the web workers will use RPyC, Execnet or similar service to access it.

**Comment by saadqc** [comment]
@agronholm Agree. But the issue is that one has to implement all this. If the feature would be available inside apscheduler package or atleast a note or a paragraph stating that apscheduler will not work with multi process web-app and one has to implement RPyC. Then a developer can be warned at the start of development instead of finding this bug on production machine.

**Comment by agronholm** [comment]
To be fair, this is not a bug of any kind. It's a deployment issue that occurs naturally not only with APScheduler but with any service that requires access to a shared resource. I'm not comfortable with adding such a thing as a "feature" to APScheduler, but perhaps it would be reasonable to document this problem and to provide a recipe to solve it.

**Comment by saadqc** [comment]
@agronholm Thanks. There are other people like me who are facing this issue on production. Now, i have to revert whole code or make significant changes in my service to make this work. I hope you understand my stance.

**Comment by shonin** [comment]
Hi @saadqc, I faced the same issue. 

I solved my problem by having my server generate a unique hash as an environment variable. The every time before I run `scheduler.start()` I check to see if the env hash is in a redis cache. If it's not, I insert it and start the scheduler, if it is, I don't start the scheduler. 

```python
def start_scheduler():
    if not IS_DEVELOPMENT:

        scheduler_instance = uuid.uuid4().hex
        app_instance = 'app__{}'.format(INSTANCE)

        if not redis_store.get(app_instance):
            redis_store.set(app_instance, scheduler_instance)
            scheduler.start()

            retrieved_scheduler_instance = redis_store.get(app_instance)
            retrieved_scheduler_instance = retrieved_scheduler_instance.decode('utf-8')

            if scheduler_instance != retrieved_scheduler_instance:
                LOGGER.info('Scheduler instance does not match cached instance ID, shutting down')
                scheduler.shutdown()

        else:
            LOGGER.info('Scheduler {} found running, duplicate scheduler not booted'.format(app_instance))
```

**Comment by seanharr11** [comment]
The two stackoverflow links posted in the first comment have been updated (Coincidentally by me, I am just reading this thread now for the first time.)

When used in any forking web server (gunicorn uwsgi, etc...), processes get forked, thus apacheduler gets instantiated (and possibly started) in each forked worker process. This leads to the "many jobs running" issue.

It is not really apschedulers responsibility to be aware of software built on top of it, that'd break abstraction.

However, as long as you call *start()* on your apscheduler object BEFORE the processes fork, and you use a consistent *jobstore* that has a single point of truth (pretty much anything but "memory"), apscheduler is capable of being run in a multi-process/worker web server environment. (b/c CRUD is performed at one location by all the workers, and jobs are only executed by the started apscheduler...)

http://stackoverflow.com/a/40162246/3923962

**Comment by saadqc** [comment]
@seanharr11 @shonin  : Agree but the non-started apscheduler will not be able to pause/resume a job. So, the request to that process (where apscheduler is not running) will not be served in that case.

**Comment by seanharr11** [comment]
I actually believe that you are still able to pause/resume jobs via any worker, as long as you are using a persistent, non-memory jobstore. I believe (correct me if I'm wrong) that the APScheduler process that is actually executing the jobs (the one you called ```start()``` on) will check the jobstore for updates to the job being run, like changes to it's state (pause/resume).

**Comment by agronholm** [comment]
It will only do this check when woken up. If the executing process is sleeping, you have to find an alternative way to wake it up after you've added a job to the job store.

**Comment by saadqc** [comment]
I worked on an alternative solution (not good but a workaround) 
Workaround:
-> Start apscheduler normally and pass a random hash in your callback method when scheduling a job
-> When job gets executed multiple time, in callback method, create an entry in your jobstore db. I used redis so I created entry with that hash (with expiry time of 4 or more seconds). this will be created when job is called first time.
-> Add a check in callback method (in start) to check for that hash entry in redis...if it is available then simply ignore other steps and return from function as we know that the callback is called multiple times.


I tested this solution with automated testing + on production it is working fine and I haven't encountered any issue yet.

**Comment by seanharr11** [comment]
I am working on support for tracking the state of jobs in the jobstore (pull request coming). 

We create a "job_submissions" collection/table/document in the jobstore, and track a submitted job throughout its lifecycle. We could then have each **Executor** check the jobstore for the job submission before redundantly running the same job over again. This feature may allow for a out-of-the-box solution for @saadqc 's concerns. 

@agronholm , are you around for a chat sometime? Once I submit the pull request (or before), I'd love to hop on a call and review with you. It's a somewhat big change, I'm just updating tests now before submitting.

**Comment by agronholm** [comment]
I just added code examples in ``examples/rpc`` via 19fa2fa27a7e168fbcec883817408a21a8b7339e. I will need to amend the documentation before I can close this issue. I hope this helps.

**Comment by matthoskins1980** [comment]
When trying the rpc example, I changed the trigger type from interval to date in client.py.  The RPC call looks like this:

```python
job = conn.root.add_job('server:print_text', 'date', args=['Hello, World'], run_date=datetime.datetime.now())
```

I get the following traceback:

> Traceback (most recent call last):
  File "client.py", line 15, in <module>
    job = conn.root.add_job('server:print_text', 'date', args=['Hello, World'], run_date=datetime.datetime.now())
  File "/home/ubuntu/.virtualenvs/yeller/local/lib/python2.7/site-packages/rpyc/core/netref.py", line 196, in __call__
    return syncreq(_self, consts.HANDLE_CALL, args, kwargs)
  File "/home/ubuntu/.virtualenvs/yeller/local/lib/python2.7/site-packages/rpyc/core/netref.py", line 71, in syncreq
    return conn.sync_request(handler, oid, *args)
  File "/home/ubuntu/.virtualenvs/yeller/local/lib/python2.7/site-packages/rpyc/core/protocol.py", line 441, in sync_request
    raise obj
AttributeError: cannot access 'tzinfo'

========= Remote Traceback (2) =========
Traceback (most recent call last):
  File "/home/ubuntu/.virtualenvs/yeller/local/lib/python2.7/site-packages/rpyc/core/protocol.py", line 305, in _dispatch_request
    res = self._HANDLERS[handler](self, *args)
  File "/home/ubuntu/.virtualenvs/yeller/local/lib/python2.7/site-packages/rpyc/core/protocol.py", line 535, in _handle_call
    return self._local_objects[oid](*args, **dict(kwargs))
  File "server.py", line 22, in exposed_add_job
    return scheduler.add_job(func, *args, **kwargs)
  File "/home/ubuntu/.virtualenvs/yeller/local/lib/python2.7/site-packages/apscheduler/schedulers/base.py", line 411, in add_job
    'trigger': self._create_trigger(trigger, trigger_args),
  File "/home/ubuntu/.virtualenvs/yeller/local/lib/python2.7/site-packages/apscheduler/schedulers/base.py", line 905, in _create_trigger
    return self._create_plugin_instance('trigger', trigger, trigger_args)
  File "/home/ubuntu/.virtualenvs/yeller/local/lib/python2.7/site-packages/apscheduler/schedulers/base.py", line 890, in _create_plugin_instance
    return plugin_cls(**constructor_kwargs)
  File "/home/ubuntu/.virtualenvs/yeller/local/lib/python2.7/site-packages/apscheduler/triggers/date.py", line 22, in __init__
    self.run_date = convert_to_datetime(run_date, timezone, 'run_date')
  File "/home/ubuntu/.virtualenvs/yeller/local/lib/python2.7/site-packages/apscheduler/util.py", line 132, in convert_to_datetime
    if datetime_.tzinfo is not None:
  File "/home/ubuntu/.virtualenvs/yeller/local/lib/python2.7/site-packages/rpyc/core/netref.py", line 150, in __getattr__
    return syncreq(self, consts.HANDLE_GETATTR, name)
  File "/home/ubuntu/.virtualenvs/yeller/local/lib/python2.7/site-packages/rpyc/core/netref.py", line 71, in syncreq
    return conn.sync_request(handler, oid, *args)
  File "/home/ubuntu/.virtualenvs/yeller/local/lib/python2.7/site-packages/rpyc/core/protocol.py", line 441, in sync_request
    raise obj
AttributeError: cannot access 'tzinfo'

========= Remote Traceback (1) =========
Traceback (most recent call last):
  File "/home/ubuntu/.virtualenvs/yeller/local/lib/python2.7/site-packages/rpyc/core/protocol.py", line 305, in _dispatch_request
    res = self._HANDLERS[handler](self, *args)
  File "/home/ubuntu/.virtualenvs/yeller/local/lib/python2.7/site-packages/rpyc/core/protocol.py", line 541, in _handle_getattr
    return self._access_attr(oid, name, (), "_rpyc_getattr", "allow_getattr", getattr)
  File "/home/ubuntu/.virtualenvs/yeller/local/lib/python2.7/site-packages/rpyc/core/protocol.py", line 504, in _access_attr
    raise AttributeError("cannot access %r" % (name,))
AttributeError: cannot access 'tzinfo'

**Comment by pierotofy** [comment]
I'll chime in to post my solution:

```python
from multiprocessing import Value

# Initialize this in the parent process
started = Value('b', False)

# child processes can call this
def start_scheduler():
    if started.value: return
    
    # initialize ap scheduler here

    started.value = True
```

**Comment by tyler46** [comment]
Hi @matthoskins1980, have you find out how to fix that issue? I am facing the same one, while trying to use trigger `date` instead of `interval`.

**Comment by agronholm** [comment]
I just opened #256 which is one solution to this problem.

**Comment by agronholm** [comment]
I have now added an entry about this in the FAQ. If you feel I should keep this issue open, please comment.

**Comment by whikwon** [comment]
@pierotofy I am using Flask + APScheduler + uWSGI and your solution doesn't work. 

Is there something I'm missing? 

**wsgi.py**
```
from multiprocessing import Value
started = Value('b', False)

from app import app

if __name__ == "__main__":
    app.run()
```

**app.py** 
```
from wsgi import started

def init_job_scheduler():
    if started.value:
        return

    scheduler = run_job_scheduler()
    started.value = True
    return scheduler

scheduler = init_job_scheduler()
```
**uwsgi.ini**
```
[uwsgi]
module=wsgi
callable=app
lazy=True
lazy-apps=False
chdir=./
http=:8086
master=True
pidfile=/tmp/slackbot.pid
vacuum=True
max-requests=5000
auto-procname=True
workers=7
enable-threads=True
```

**Comment by Laurel-rao** [comment]
Is that distribute feature be come true?
